File: packages/asdl-slots/src/asdl_slots/lifecycle/pool.py
"""Pool initialization and resizing for slot lifecycle."""

from __future__ import annotations

from dataclasses import dataclass

from asdl_slots.context import SlotsCliContext
from asdl_slots.inventory import SlotInventory, SlotRecord, build_slot_inventory
from asdl_slots.lifecycle.operation_state import slot_operation_in_progress_message
from asdl_slots.lifecycle.outcomes import (
    SlotInitOutcome,
    SlotLifecycleFailure,
    SlotResizeOutcome,
)
from asdl_slots.naming import generate_slot_name
from asdl_slots.repo_context import ensure_slots_metadata_dir

MIN_POOL_SIZE = 1
MAX_POOL_SIZE = 99


@dataclass(frozen=True)
class InitPlan:
    create: tuple[int, ...]


@dataclass(frozen=True)
class ResizePlan:
    create: tuple[int, ...]
    remove: tuple[SlotRecord, ...]


def _invalid_size_failure() -> SlotLifecycleFailure:
    return SlotLifecycleFailure(
        error_type="invalid_size",
        message=f"--size must be between {MIN_POOL_SIZE} and {MAX_POOL_SIZE}.",
    )


def build_init_plan(target_size: int) -> InitPlan:
    return InitPlan(create=tuple(range(1, target_size + 1)))


def build_resize_plan(inventory: SlotInventory, target_size: int) -> ResizePlan:
    if target_size == inventory.pool_size:
        return ResizePlan(create=(), remove=())
    if target_size > inventory.pool_size:
        existing = {record.slot_number for record in inventory.records}
        needed = target_size - inventory.pool_size
        create: list[int] = []
        candidate = 1
        while len(create) < needed:
            if candidate not in existing:
                create.append(candidate)
            candidate += 1
        return ResizePlan(create=tuple(create), remove=())
    sorted_records = sorted(inventory.records, key=lambda r: r.slot_number)
    return ResizePlan(create=(), remove=tuple(sorted_records[target_size:]))


def initialize_pool(
    slots_ctx: SlotsCliContext, target_size: int
) -> SlotInitOutcome | SlotLifecycleFailure:
    if target_size < MIN_POOL_SIZE or target_size > MAX_POOL_SIZE:
        return _invalid_size_failure()

    inventory = build_slot_inventory(
        slots_ctx.git,
        main_repo_root=slots_ctx.repo.main_repo_root,
    )
    if inventory.pool_size > 0:
        return SlotLifecycleFailure(
            error_type="pool_already_initialized",
            message=(
                f"Pool already has {inventory.pool_size} slot(s). "
                f"Use `slot resize --size N` to change capacity."
            ),
        )

    ensure_slots_metadata_dir(slots_ctx.repo, slots_ctx.storage)
    plan = build_init_plan(target_size)
    trunk = slots_ctx.git.get_trunk_branch()
    created: list[str] = []
    for slot_number in plan.create:
        name = generate_slot_name(slot_number)
        path = slots_ctx.repo.worktrees_dir / name
        slots_ctx.git.add_detached_worktree(path, trunk)
        created.append(name)

    return SlotInitOutcome(
        created=tuple(created),
        pool_size=len(created),
        worktrees_dir=slots_ctx.repo.worktrees_dir,
    )


def resize_pool(
    slots_ctx: SlotsCliContext, target_size: int
) -> SlotResizeOutcome | SlotLifecycleFailure:
    if target_size < MIN_POOL_SIZE or target_size > MAX_POOL_SIZE:
        return _invalid_size_failure()

    inventory = build_slot_inventory(
        slots_ctx.git,
        main_repo_root=slots_ctx.repo.main_repo_root,
    )
    previous_pool_size = inventory.pool_size
    plan = build_resize_plan(inventory, target_size)

    if not plan.create and not plan.remove:
        return SlotResizeOutcome(
            previous_pool_size=previous_pool_size,
            pool_size=previous_pool_size,
            created=(),
            removed=(),
            worktrees_dir=slots_ctx.repo.worktrees_dir,
        )

    if plan.remove:
        errors = _validate_removals(slots_ctx, plan.remove)
        if errors:
            return SlotLifecycleFailure(
                error_type="resize_unsafe",
                message="\n".join(errors),
            )

    ensure_slots_metadata_dir(slots_ctx.repo, slots_ctx.storage)
    trunk = slots_ctx.git.get_trunk_branch()

    created: list[str] = []
    for slot_number in plan.create:
        name = generate_slot_name(slot_number)
        path = slots_ctx.repo.worktrees_dir / name
        slots_ctx.git.add_detached_worktree(path, trunk)
        created.append(name)

    removed: list[str] = []
    for record in plan.remove:
        slots_ctx.git.remove_worktree(record.path)
        removed.append(record.slot_name)

    return SlotResizeOutcome(
        previous_pool_size=previous_pool_size,
        pool_size=previous_pool_size + len(created) - len(removed),
        created=tuple(created),
        removed=tuple(removed),
        worktrees_dir=slots_ctx.repo.worktrees_dir,
    )


def _validate_removals(
    slots_ctx: SlotsCliContext, to_remove: tuple[SlotRecord, ...]
) -> tuple[str, ...]:
    errors: list[str] = []
    for record in to_remove:
        if record.operation is not None:
            errors.append(
                slot_operation_in_progress_message(
                    slot_name=record.slot_name,
                    branch_name=record.branch,
                    worktree_path=record.path,
                    operation=record.operation,
                    action="shrinking the pool",
                )
            )
            continue
        if record.branch is not None:
            errors.append(
                f"{record.slot_name} is assigned to '{record.branch}'; "
                f"free it before shrinking the pool."
            )
            continue
        if slots_ctx.git.has_uncommitted_changes(record.path):
            errors.append(
                f"{record.slot_name} at {record.path} has uncommitted changes; "
                f"commit or discard before shrinking the pool."
            )
    return tuple(errors)

File: ts/packages/slot/src/lifecycle/pool.ts
import { join } from "node:path";

import { buildSlotInventory, poolSize, type SlotInventory, type SlotRecord } from "../inventory.ts";
import { generateSlotName } from "../naming.ts";
import { ensureSlotsMetadataDir } from "../repo-context.ts";
import type { SlotCliContext } from "../context.ts";

export const MIN_POOL_SIZE = 1;
export const MAX_POOL_SIZE = 99;

export interface InitPlan {
	create: readonly number[];
}

export interface ResizePlan {
	create: readonly number[];
	remove: readonly SlotRecord[];
}

export interface SlotInitOutcome {
	created: string[];
	pool_size: number;
	worktrees_dir: string;
}

export interface SlotResizeOutcome {
	previous_pool_size: number;
	pool_size: number;
	created: string[];
	removed: string[];
	worktrees_dir: string;
}

export interface SlotLifecycleFailure {
	error_type: string;
	message: string;
}

export type SlotInitResult = { type: "ok"; outcome: SlotInitOutcome } | { type: "failure"; failure: SlotLifecycleFailure };
export type SlotResizeResult = { type: "ok"; outcome: SlotResizeOutcome } | { type: "failure"; failure: SlotLifecycleFailure };

export function buildInitPlan(targetSize: number): InitPlan {
	return { create: Array.from({ length: targetSize }, (_value, index) => index + 1) };
}

export function buildResizePlan(inventory: SlotInventory, targetSize: number): ResizePlan {
	const currentSize = poolSize(inventory);
	if (targetSize === currentSize) return { create: [], remove: [] };
	if (targetSize > currentSize) {
		const existing = new Set(inventory.records.map((record) => record.slotNumber));
		const needed = targetSize - currentSize;
		const create: number[] = [];
		let candidate = 1;
		while (create.length < needed) {
			if (!existing.has(candidate)) create.push(candidate);
			candidate += 1;
		}
		return { create, remove: [] };
	}
	const sortedRecords = [...inventory.records].sort((left, right) => left.slotNumber - right.slotNumber);
	return { create: [], remove: sortedRecords.slice(targetSize) };
}

export async function initializePool(ctx: SlotCliContext, targetSize: number): Promise<SlotInitResult> {
	if (ctx.repo.type !== "repo") return { type: "failure", failure: { error_type: ctx.repo.errorType, message: ctx.repo.message } };
	const sizeFailure = invalidSizeFailure(targetSize);
	if (sizeFailure !== null) return { type: "failure", failure: sizeFailure };
	const inventory = await buildSlotInventory(ctx.git, { mainRepoRoot: ctx.repo.mainRepoRoot });
	if (poolSize(inventory) > 0) {
		return {
			type: "failure",
			failure: {
				error_type: "pool_already_initialized",
				message: `Pool already has ${poolSize(inventory)} slot(s). Use \`slot resize --size N\` to change capacity.`,
			},
		};
	}
	await ensureSlotsMetadataDir(ctx.repo, ctx.storage);
	const trunk = await ctx.git.getTrunkBranch();
	const created: string[] = [];
	for (const slotNumber of buildInitPlan(targetSize).create) {
		const name = generateSlotName(slotNumber);
		await ctx.git.addDetachedWorktree(join(ctx.repo.worktreesDir, name), trunk);
		created.push(name);
	}
	return { type: "ok", outcome: { created, pool_size: created.length, worktrees_dir: ctx.repo.worktreesDir } };
}

export async function resizePool(ctx: SlotCliContext, targetSize: number): Promise<SlotResizeResult> {
	if (ctx.repo.type !== "repo") return { type: "failure", failure: { error_type: ctx.repo.errorType, message: ctx.repo.message } };
	const sizeFailure = invalidSizeFailure(targetSize);
	if (sizeFailure !== null) return { type: "failure", failure: sizeFailure };
	const inventory = await buildSlotInventory(ctx.git, { mainRepoRoot: ctx.repo.mainRepoRoot });
	const previousPoolSize = poolSize(inventory);
	const plan = buildResizePlan(inventory, targetSize);
	if (plan.create.length === 0 && plan.remove.length === 0) {
		return {
			type: "ok",
			outcome: { previous_pool_size: previousPoolSize, pool_size: previousPoolSize, created: [], removed: [], worktrees_dir: ctx.repo.worktreesDir },
		};
	}
	if (plan.remove.length > 0) {
		const errors = await validateRemovals(ctx, plan.remove);
		if (errors.length > 0) return { type: "failure", failure: { error_type: "resize_unsafe", message: errors.join("\n") } };
	}
	await ensureSlotsMetadataDir(ctx.repo, ctx.storage);
	const trunk = await ctx.git.getTrunkBranch();
	const created: string[] = [];
	for (const slotNumber of plan.create) {
		const name = generateSlotName(slotNumber);
		await ctx.git.addDetachedWorktree(join(ctx.repo.worktreesDir, name), trunk);
		created.push(name);
	}
	const removed: string[] = [];
	for (const record of plan.remove) {
		await ctx.git.removeWorktree(record.path);
		removed.push(record.slotName);
	}
	return {
		type: "ok",
		outcome: {
			previous_pool_size: previousPoolSize,
			pool_size: previousPoolSize + created.length - removed.length,
			created,
			removed,
			worktrees_dir: ctx.repo.worktreesDir,
		},
	};
}

async function validateRemovals(ctx: SlotCliContext, records: readonly SlotRecord[]): Promise<readonly string[]> {
	const errors: string[] = [];
	for (const record of records) {
		if (record.operation !== null) {
			errors.push(slotOperationInProgressMessage(record));
			continue;
		}
		if (record.branch !== null) {
			errors.push(`${record.slotName} is assigned to '${record.branch}'; free it before shrinking the pool.`);
			continue;
		}
		if (await ctx.git.hasUncommittedChanges(record.path)) {
			errors.push(`${record.slotName} at ${record.path} has uncommitted changes; commit or discard before shrinking the pool.`);
		}
	}
	return errors;
}

function slotOperationInProgressMessage(record: SlotRecord): string {
	const branchText = record.branch === null ? "unknown branch" : `'${record.branch}'`;
	return `${record.slotName} is assigned to ${branchText} at ${record.path} with ${record.operation} in progress; finish or abort it before shrinking the pool.`;
}

function invalidSizeFailure(targetSize: number): SlotLifecycleFailure | null {
	if (Number.isInteger(targetSize) && targetSize >= MIN_POOL_SIZE && targetSize <= MAX_POOL_SIZE) return null;
	return { error_type: "invalid_size", message: `--size must be between ${MIN_POOL_SIZE} and ${MAX_POOL_SIZE}.` };
}
